File: tests/benchmark.py
import asyncio
import time
import random

async def client_session(client_id: int, num_messages: int, host: str = '127.0.0.1', port: int = 8080):
    for i in range(num_messages):
        try:
            reader, writer = await asyncio.open_connection(host, port)
            message = f"Message {i} from client {client_id}\n"
            writer.write(message.encode())
            await writer.drain()

            data = await reader.read(1024)
            writer.close()
            await writer.wait_closed()

            await asyncio.sleep(random.uniform(0.1, 0.5))

        except Exception as e:
            print(f"Client {client_id} error: {e}")

async def benchmark(num_clients: int, messages_per_client: int):
    tasks = [client_session(i, messages_per_client) for i in range(num_clients)]
    start_time = time.time()
    await asyncio.gather(*tasks)
    end_time = time.time()

    total_messages = num_clients * messages_per_client
    total_time = end_time - start_time
    qps = total_messages / total_time

    print(f"Total messages: {total_messages}")
    print(f"Total time: {total_time:.2f} seconds")
    print(f"QPS: {qps:.2f}")

if __name__ == "__main__":
    num_clients = 100
    messages_per_client = 100
    asyncio.run(benchmark(num_clients, messages_per_client))

File: tests/benchmark.go
package main

import (
	"fmt"
	"io"
	"math/rand"
	"net"
	"sync"
	"time"
)

func clientSession(clientID, numMessages int, host string, port int, wg *sync.WaitGroup, results chan<- [2]int) {
	defer wg.Done()
	totalBytesSent := 0
	totalBytesReceived := 0

	for i := 0; i < numMessages; i++ {
		conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", host, port))
		if err != nil {
			fmt.Printf("Client %d error: %v\n", clientID, err)
			return
		}

		message := fmt.Sprintf("Message %d from client %d\n", i, clientID)
		fmt.Printf("Client %d sending: %s", clientID, message)

		n, err := conn.Write([]byte(message))
		if err != nil {
			fmt.Printf("Client %d error: %v\n", clientID, err)
			conn.Close()
			return
		}
		totalBytesSent += n

		buf := make([]byte, 1024)
		n, err = conn.Read(buf)
		if err != nil && err != io.EOF {
			fmt.Printf("Client %d error: %v\n", clientID, err)
			conn.Close()
			return
		}
		totalBytesReceived += n

		fmt.Printf("Client %d received: %s", clientID, string(buf[:n]))

		conn.Close()

		// Random delay between messages
		time.Sleep(time.Duration(rand.Float64()*0.4+0.1) * time.Second)
	}

	results <- [2]int{totalBytesSent, totalBytesReceived}
}

func main() {
	numClients := 1000
	messagesPerClient := 100
	host := "127.0.0.1"
	port := 8080

	var wg sync.WaitGroup
	results := make(chan [2]int, numClients)

	startTime := time.Now()

	for i := 0; i < numClients; i++ {
		wg.Add(1)
		go clientSession(i, messagesPerClient, host, port, &wg, results)
	}

	wg.Wait()
	close(results)

	totalBytesSent := 0
	totalBytesReceived := 0
	for result := range results {
		totalBytesSent += result[0]
		totalBytesReceived += result[1]
	}

	totalMessages := numClients * messagesPerClient
	totalTime := time.Since(startTime).Seconds()
	qps := float64(totalMessages) / totalTime
	totalBytes := totalBytesSent + totalBytesReceived
	bandwidth := float64(totalBytes) / totalTime / (1024 * 1024) // Convert to MB/s

	fmt.Printf("Total messages: %d\n", totalMessages)
	fmt.Printf("Total time: %.2f seconds\n", totalTime)
	fmt.Printf("QPS: %.2f\n", qps)
	fmt.Printf("Total bytes sent: %d\n", totalBytesSent)
	fmt.Printf("Total bytes received: %d\n", totalBytesReceived)
	fmt.Printf("Bandwidth: %.2f MB/s\n", bandwidth)
}